### ISSUE TITLE
XFS recovery mount triggered on boot after using a run podman volume

### ISSUE THREAD

**Issue by basvdlei** [coreos/fedora-coreos-tracker#1178]
**Describe the bug**
The sysroot xfs filesystem is mounted with recovery, instead of cleanly, when a podman container with a `-v /run:/run` was started.

**Reproduction steps**
1. Run a container with root podman with a `/run:/run` volume: `sudo /bin/podman run --volume /run:/run --entrypoint=/bin/sleep registry.fedoraproject.org/fedora-minimal:35 1`
2. Reboot the node `sudo systemctl reboot`
3. Observe the boot logs `sudo journalctl -b | grep XFS`

**Expected behavior**
See that the filesystem is mounted cleanly:

```
localhost kernel: SGI XFS with ACLs, security attributes, scrub, quota, no debug enabled
localhost kernel: XFS (vda4): Mounting V5 Filesystem
localhost kernel: XFS (vda4): Ending clean mount
localhost kernel: xfs filesystem being mounted at /sysroot supports timestamps until 2038 (0x7fffffff)
localhost systemd[1]: Mounted /sysroot.
```

**Actual behavior**
During the mount recovery is initiated:
```
localhost kernel: SGI XFS with ACLs, security attributes, scrub, quota, no debug enabled
localhost kernel: XFS (vda4): Mounting V5 Filesystem
localhost kernel: XFS (vda4): Starting recovery (logdev: internal)
localhost kernel: XFS (vda4): Ending recovery (logdev: internal)
localhost kernel: xfs filesystem being mounted at /sysroot supports timestamps until 2038 (0x7fffffff)
localhost systemd[1]: Mounted /sysroot.
```

**System details**
 - QEMU
 - 35.20220327.3.0 (current stable)
 - 36.20220418.1.0 (current next)

**Ignition config**
```
variant: fcos
version: 1.1.0

passwd:
  users:
    - name: core
      ssh_authorized_keys:
        - REDACTED

systemd:
  units:
    - name: "rpm-ostree-countme.timer"
      mask: true
    - name: "rpm-ostree-countme.service"
      mask: true
    - name: "fstrim.timer"
      mask: true
    - name: "fstrim.service"
      mask: true
    - name: "zincati.service"
      mask: true
    - name: "xfskiller.service"
      enabled: true
      contents: |
        [Unit]
        Description=XFS corruption testcase
        After=network-online.target
        Wants=network-online.target
        [Service]
        Type=oneshot
        ExecStartPre=-/bin/podman kill xfskiller
        ExecStartPre=-/bin/podman rm xfskiller
        ExecStart=/bin/podman run \
          --name xfskiller \
          --volume /run:/run \
          --entrypoint=/bin/sleep \
          registry.fedoraproject.org/fedora-minimal:35 1
        ExecStop=-/bin/podman stop xfskiller
        [Install]
        WantedBy=multi-user.target
```

**Additional information**
We've observed an increasing amount of nodes that can't reboot due to a corrupt sysroot filesystem. While initially blaming unreliable storage, it turned out that all our nodes are mounting the root filesystem in recovery mode due the issue described above.

We run a Kubelet inside a podman container, similar to [Typhoon](https://github.com/poseidon/typhoon). It also has this volume mount: https://github.com/poseidon/typhoon/blob/d7f55c4e46b6c6883edb14b6954499949caa0a24/bare-metal/fedora-coreos/kubernetes/fcc/worker.yaml#L45

My testing indicates that this issue has been introduced when Fedora CoreOS rebased from Fedora 33 to 34, since I cannot reproduce it on 33.20210426.3.0, but reliably on 34.20210324.10.0.

Mounting `/run` on a different location inside the container does not seem to trigger this issue. I'm testing a work-around by using a `-v /run:/host/run` volume and bind mounting this to `/run` inside the container.

**Comment by basvdlei** [comment]
My testing confirms that bind mounting from inside the container does not trigger this issue.

```
sudo podman run --rm --privileged --volume /run:/host/run \
     registry.fedoraproject.org/fedora-minimal:35 bash -c \
         "microdnf install -y util-linux-core && mount -o bind /host/run /run && sleep 1"
```

Out of 45 nodes in my setup that did not yet have this workaround in place, 20% were broken on reboot due to a corrupted XFS sysroot. While 150 nodes with the workaround in place, all rebooted successfully.

**Comment by cmurf** [comment]
I'm confused. `/run` should be a tmpfs. So why does this result in an XFS file system being mounted? And then I wonder if this same XFS file system is already mounted before the podman command which results in the indication of journal replay.

As far as I'm aware, XFS shouldn't allow multiple mounts of the same file system. So maybe the container is preventing the normal check that warns and stops mounting an already mounted XFS? This would explain why the mount attempt shows journal recovery, but the bind mount doesn't.

**Comment by basvdlei** [comment]
Agreed, I've been banging my head on this as well. `/run` is a tmpfs, and by creating a volume with podman, crun is doing a bind mount as well. I'm not sure if this issue is caused by the filesystem being mounted again or if somehow some data is written to the disk.

It might have something to do with `/run` being there while podman is doing some additional setup like creating the secrets, etc. This is the contents of `/run` in a podman container:
```
bash-5.1# ls -la /run/
total 0
drwxr-xr-x. 1 root root 42 May  9 14:44 .
dr-xr-xr-x. 1 root root 17 May  9 14:44 ..
-rw-r--r--. 1 root root  0 May  9 14:44 .containerenv
drwxr-xr-x. 2 root root  6 Feb 21 06:48 lock
drwxr-xr-x. 2 root root 40 May  9 14:44 secrets
```

**Comment by cgwalters** [comment]
I think looking at `journalctl -b -1` i.e. viewing the shutdown process would likely show that there's some sort of ordering problem during shutdown.  We have some outstanding bits related to that in ostree core that we need to fix.  I suspect the container isn't being terminated early enough, so its mount for `/run` is holding things up, and preventing unclean mount of `/` aka `/sysroot`.

**Comment by basvdlei** [comment]
This issue is fixed in recent releases of Fedora CoreOS.

I went back through old testing stream releases to figure out when this was fixed. `36.20220806.2.0`  was the last FCOS release that still had this issue. While an update to `36.20220820.2.0` fixed it. I downgraded packages one-by-one from the release diff and I was able to reintroduced the issue by downgrading podman.

This issue is fixed by the Podman update from `4:4.1.1-3.fc36` to `4:4.2.0-2.fc36`.
